File: src/Auth/Email.hs
module Auth.Email
  ( sendVerificationEmail
  ) where

import           Data.Text          (Text)
import qualified Data.Text          as T
import qualified Data.Text.Lazy     as TL
import           Network.Mail.Mime  (Address (..), simpleMail')
import           Network.Mail.SMTP  (sendMailWithLoginTLS')

import           Auth.Config        (Config (..), SmtpConfig (..))

sendVerificationEmail :: Config -> Text -> Text -> IO ()
sendVerificationEmail cfg toEmail link =
  case cfgSmtp cfg of
    Nothing ->
      putStrLn $ "[dev-email] verification link for " <> T.unpack toEmail
                 <> ":\n            " <> T.unpack link
    Just smtp -> do
      let from    = Address (Just (cfgRpName cfg)) (smtpFrom smtp)
          to      = Address Nothing toEmail
          subject = "Verify your email address"
          body    = TL.fromStrict $
                      "Welcome to " <> cfgRpName cfg <> "!\n\n"
                      <> "Please verify your email address by opening:\n\n"
                      <> link <> "\n\n"
                      <> "If you did not create an account you can ignore this message.\n"
          mail    = simpleMail' to from subject body
      sendMailWithLoginTLS'
        (smtpHost smtp)
        (fromIntegral (smtpPort smtp))
        (smtpUser smtp)
        (smtpPass smtp)
        mail

File: src/auth/email.rs
use anyhow::Context;
use lettre::message::Mailbox;
use lettre::transport::smtp::authentication::Credentials;
use lettre::{AsyncSmtpTransport, AsyncTransport, Message, Tokio1Executor};

use crate::config::Config;

pub async fn send_verification_email(config: &Config, to_email: &str, link: &str) {
    match &config.smtp {
        None => {
            println!(
                "[dev-email] verification link for {to_email}:\n            {link}"
            );
        }
        Some(_) => {
            if let Err(err) = send_via_smtp(config, to_email, link).await {
                tracing::error!(error = %err, "failed to send verification email");
            }
        }
    }
}

async fn send_via_smtp(config: &Config, to_email: &str, link: &str) -> anyhow::Result<()> {
    let smtp = config
        .smtp
        .as_ref()
        .context("smtp configuration missing")?;

    let from = Mailbox::new(
        Some(config.rp_name.clone()),
        smtp.from.parse().context("invalid SMTP_FROM address")?,
    );
    let to = Mailbox::new(None, to_email.parse().context("invalid recipient address")?);
    let body = format!(
        "Welcome to {name}!\n\n\
         Please verify your email address by opening:\n\n\
         {link}\n\n\
         If you did not create an account you can ignore this message.\n",
        name = config.rp_name,
    );

    let email = Message::builder()
        .from(from)
        .to(to)
        .subject("Verify your email address")
        .body(body)
        .context("building email")?;

    let mailer = AsyncSmtpTransport::<Tokio1Executor>::relay(&smtp.host)
        .context("configuring SMTP transport")?
        .port(smtp.port)
        .credentials(Credentials::new(smtp.user.clone(), smtp.pass.clone()))
        .build();

    mailer.send(email).await.context("sending email")?;
    Ok(())
}
